### PR TITLE
Enable automatically using import hook and env variable

### PR DESCRIPTION
**Pull request by baldassarreFe**
# Problem

Right now, users need to import lovely tensors and monkey patch:
```python
import lovely_tensors
lovely_tensors.monkey_patch()
```

Doing this in every notebook, training script, evaluation script, python console, etc. can become repetitive.

# Solution

This PR allows enabling lovely tensors automatically without any modification to the code. All is needed is that the pip package is installed and that a variable `LOVELY_TENSORS=1` is set in the environment:
```bash
python -c 'import torch; print(torch.ones(3))'
# tensor([1., 1., 1.])

LOVELY_TENSORS=1 python -c 'import torch; print(torch.ones(3))'
# tensor[3] x∈[1.000, 1.000] μ=1.000 σ=0. [1.000, 1.000, 1.000]
```

# Notes

Adding the variable to the environment can be done in multiple ways:
- inline just before calling python, as shown above
- for a single terminal session, eg `export echo LOVELY_TENSORS=1`
- for all terminal sessions by modifying `.bashrc`, eg `echo LOVELY_TENSORS=1 >> ~/.bashrc`
- for a specific conda environment eg `conda env config vars set LOVELY_TENSORS=1`

If the env variable is not set, or is set to 0 or an empty string, the import hook does nothing. In this case, the user can still follow the old workflow or importing and monkey patching, this PR does not modify that behavior.

# References

This way of enabling a feature using only an env variable and without explicit imports is inspired by the awesome package [better exceptions](https://github.com/Qix-/better-exceptions). 

## Summary by CodeRabbit

- **New Features**
	- Introduced a new hook functionality for the `lovely_tensors` library, activated via an environment variable.
	- Added a new file, `lovely_tensors_hook.pth`, to the package distribution.

- **Enhancements**
	- Improved setup process with new classes and functions to manage the hook file during installation and build processes.

- **Bug Fixes**
	- Enhanced error handling in the hooking process for better user guidance.

### CODE REVIEW

**Comment by baldassarreFe** [comment]
Before merging we should also document the feature in the README and maybe bump the version.

**Comment by xl0** [comment]
Wow @coderabbitai got really annoying, I'll disable it for now.

@baldassarreFe Thank you, I like this idea. Let me test it a bit, and I'll merge it.

**Comment by baldassarreFe** [comment]
> Wow https://github.com/coderabbitai got really annoying, I'll disable it for now.

Well, it's very verbose but it's not wrong, mostly :D

- I'll skip the suggestion about logging because at that point logging is probably not init yet.
- The suggestion about `setuptools.command.build_py` makes sense but I could not get it to work, I'll keep using `distutils` for now, hopefully someone more knowledgeable will pitch in
- The suggestion about itertools and the list is good
- The suggestion about removing the exec will not work because the hook file needs to be on a single line (not sure why but it doesn't run if it's multiple lines)

> Thank you, I like this idea. Let me test it a bit, and I'll merge it.

Great! I'm glad you liked it

File: setup.py [hunk 1]
```diff
@@ -1,8 +1,61 @@
...
+class InstallLibWithHook(install_lib):
+    def run(self):
+        install_lib.run(self)
+        path = os.path.join(self.build_dir, LOVELY_TENSORS_HOOK)
+        dest = os.path.join(self.install_dir, LOVELY_TENSORS_HOOK)
+        self.copy_file(path, dest)
+        self.outputs = [dest]
+
+    def get_outputs(self):
+        return itertools.chain(install_lib.get_outputs(self), self.outputs)
```
[the remark this answers is not in the document]
Comment (baldassarreFe) [reply setup.py]: Ok 👍

File: lovely_tensors_hook.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ERROR_MESSAGE = """\
+An error occurred while automatically hooking lovely_tensors.
+If you uninstalled lovely_tensors, you should probably delete any 'lovely_tensors_hook.pth'
+file on your system or unset your 'LOVELY_TENSORS' environment variable.
+"""
+
+try:
+    if os.environ.get("LOVELY_TENSORS", "") not in {"", "0"}:
+        import lovely_tensors
+        lovely_tensors.monkey_patch()
```
[the remark this answers is not in the document]
Comment (baldassarreFe) [reply lovely_tensors_hook.py:12]: Documentation is definitely needed, but is not the topic of this PR

File: lovely_tensors_hook.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+except Exception as e:
+    print(ERROR_MESSAGE)
+    print(e)
+    raise
```
[the remark this answers is not in the document]
Comment (baldassarreFe) [reply lovely_tensors_hook.py:16]: Logging is not yet initialized at this stage

**Comment by xl0** [comment]
@baldassarreFe I gave it a go. 2 comments

1. `pip install -e .` fails on python=3.12.5. This is the normal way to do development with nbdev, as we need to keep the generated python files in sync with what other parts of the library import. I have not looked into it, but I think a bunch of stuff got deprecated lately.

2. How do you feel about moving/adding the check for LOVELY_TENSORS to __init__.py ? This way, I can set it in .env and import lovely_tensors, I will activate automatically.

The error:
```
$ pip install -e .
Obtaining file:///home/xl0/work/projects/lovely-tensors
  Preparing metadata (setup.py) ... done
Requirement already satisfied: torch in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-tensors==0.1.17) (2.4.0)
Requirement already satisfied: lovely-numpy>=0.2.13 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-tensors==0.1.17) (0.2.13)
Requirement already satisfied: numpy>=1.17 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (2.1.0)
Requirement already satisfied: fastcore in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (1.7.3)
Requirement already satisfied: ipython in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (8.27.0)
Requirement already satisfied: matplotlib in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (3.9.2)
Requirement already satisfied: filelock in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (3.15.4)
Requirement already satisfied: typing-extensions>=4.8.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (4.12.2)
Requirement already satisfied: sympy in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (1.13.2)
Requirement already satisfied: networkx in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (3.3)
Requirement already satisfied: jinja2 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (3.1.4)
Requirement already satisfied: fsspec in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (2024.6.1)
Requirement already satisfied: setuptools in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from torch->lovely-tensors==0.1.17) (72.2.0)
Requirement already satisfied: packaging in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from fastcore->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (24.1)
Requirement already satisfied: decorator in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (5.1.1)
Requirement already satisfied: jedi>=0.16 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.19.1)
Requirement already satisfied: matplotlib-inline in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.1.7)
Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (3.0.47)
Requirement already satisfied: pygments>=2.4.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (2.18.0)
Requirement already satisfied: stack-data in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.6.2)
Requirement already satisfied: traitlets>=5.13.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (5.14.3)
Requirement already satisfied: pexpect>4.3 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (4.9.0)
Requirement already satisfied: MarkupSafe>=2.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from jinja2->torch->lovely-tensors==0.1.17) (2.1.5)
Requirement already satisfied: contourpy>=1.0.1 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (1.3.0)
Requirement already satisfied: cycler>=0.10 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.12.1)
Requirement already satisfied: fonttools>=4.22.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (4.53.1)
Requirement already satisfied: kiwisolver>=1.3.1 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (1.4.5)
Requirement already satisfied: pillow>=8 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (10.4.0)
Requirement already satisfied: pyparsing>=2.3.1 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (3.1.4)
Requirement already satisfied: python-dateutil>=2.7 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (2.9.0)
Requirement already satisfied: mpmath<1.4,>=1.1.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from sympy->torch->lovely-tensors==0.1.17) (1.3.0)
Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from jedi>=0.16->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.8.4)
Requirement already satisfied: ptyprocess>=0.5 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from pexpect>4.3->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.7.0)
Requirement already satisfied: wcwidth in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from prompt-toolkit<3.1.0,>=3.0.41->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.2.13)
Requirement already satisfied: six>=1.5 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from python-dateutil>=2.7->matplotlib->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (1.16.0)
Requirement already satisfied: executing>=1.2.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from stack-data->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (2.0.1)
Requirement already satisfied: asttokens>=2.1.0 in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from stack-data->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (2.4.1)
Requirement already satisfied: pure-eval in /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages (from stack-data->ipython->lovely-numpy>=0.2.13->lovely-tensors==0.1.17) (0.2.3)
Installing collected packages: lovely-tensors
  DEPRECATION: Legacy editable install of lovely-tensors==0.1.17 from file:///home/xl0/work/projects/lovely-tensors (setup.py develop) is deprecated. pip 25.0 will enforce this behaviour change. A possible replacement is to add a pyproject.toml or enable --use-pep517, and use setuptools >= 64. If the resulting installation is not behaving as expected, try using --config-settings editable_mode=compat. Please consult the setuptools documentation for more information. Discussion can be found at https://github.com/pypa/pip/issues/11457
  Running setup.py develop for lovely-tensors
    error: subprocess-exited-with-error
    
    × python setup.py develop did not run successfully.
    │ exit code: 1
    ╰─> [71 lines of output]
        /home/xl0/work/projects/lovely-tensors/setup.py:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
          from pkg_resources import parse_version
        running develop
        /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/command/develop.py:42: EasyInstallDeprecationWarning: easy_install command is deprecated.
        !!
        
                ********************************************************************************
                Please avoid running ``setup.py`` and ``easy_install``.
                Instead, use pypa/build, pypa/installer or other
                standards-based tools.
        
                See https://github.com/pypa/setuptools/issues/917 for details.
                ********************************************************************************
        
        !!
          easy_install.initialize_options(self)
        /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
        !!
        
                ********************************************************************************
                Please avoid running ``setup.py`` directly.
                Instead, use pypa/build, pypa/installer or other
                standards-based tools.
        
                See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
                ********************************************************************************
        
        !!
          self.initialize_options()
        running egg_info
        writing lovely_tensors.egg-info/PKG-INFO
        writing dependency_links to lovely_tensors.egg-info/dependency_links.txt
        writing entry points to lovely_tensors.egg-info/entry_points.txt
        writing requirements to lovely_tensors.egg-info/requires.txt
        writing top-level names to lovely_tensors.egg-info/top_level.txt
        reading manifest file 'lovely_tensors.egg-info/SOURCES.txt'
        reading manifest template 'MANIFEST.in'
        warning: no files found matching 'CONTRIBUTING.md'
        warning: no files found matching 'lovely_tensors_hook.pth'
        warning: no previously-included files matching '__pycache__' found under directory '*'
        adding license file 'LICENSE'
        writing manifest file 'lovely_tensors.egg-info/SOURCES.txt'
        running build_ext
        Creating /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/lovely-tensors.egg-link (link to .)
        Adding lovely-tensors 0.1.17 to easy-install.pth file
        
        Installed /home/xl0/work/projects/lovely-tensors
        Traceback (most recent call last):
          File "<string>", line 2, in <module>
          File "<pip-setuptools-caller>", line 34, in <module>
          File "/home/xl0/work/projects/lovely-tensors/setup.py", line 85, in <module>
            setuptools.setup(
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/__init__.py", line 108, in setup
            return distutils.core.setup(**attrs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 184, in setup
            return run_commands(dist)
                   ^^^^^^^^^^^^^^^^^^
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 200, in run_commands
            dist.run_commands()
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 964, in run_commands
            self.run_command(cmd)
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/dist.py", line 945, in run_command
            super().run_command(command)
          File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 983, in run_command
            cmd_obj.run()
          File "/home/xl0/work/projects/lovely-tensors/setup.py", line 53, in run
            path = os.path.join(self.build_directory, LOVELY_TENSORS_HOOK)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "<frozen posixpath>", line 76, in join
        TypeError: expected str, bytes or os.PathLike object, not NoneType
        [end of output]
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× python setup.py develop did not run successfully.
│ exit code: 1
╰─> [71 lines of output]
    /home/xl0/work/projects/lovely-tensors/setup.py:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
      from pkg_resources import parse_version
    running develop
    /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/command/develop.py:42: EasyInstallDeprecationWarning: easy_install command is deprecated.
    !!
    
            ********************************************************************************
            Please avoid running ``setup.py`` and ``easy_install``.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.
    
            See https://github.com/pypa/setuptools/issues/917 for details.
            ********************************************************************************
    
    !!
      easy_install.initialize_options(self)
    /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
    !!
    
            ********************************************************************************
            Please avoid running ``setup.py`` directly.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.
    
            See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
            ********************************************************************************
    
    !!
      self.initialize_options()
    running egg_info
    writing lovely_tensors.egg-info/PKG-INFO
    writing dependency_links to lovely_tensors.egg-info/dependency_links.txt
    writing entry points to lovely_tensors.egg-info/entry_points.txt
    writing requirements to lovely_tensors.egg-info/requires.txt
    writing top-level names to lovely_tensors.egg-info/top_level.txt
    reading manifest file 'lovely_tensors.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching 'CONTRIBUTING.md'
    warning: no files found matching 'lovely_tensors_hook.pth'
    warning: no previously-included files matching '__pycache__' found under directory '*'
    adding license file 'LICENSE'
    writing manifest file 'lovely_tensors.egg-info/SOURCES.txt'
    running build_ext
    Creating /home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/lovely-tensors.egg-link (link to .)
    Adding lovely-tensors 0.1.17 to easy-install.pth file
    
    Installed /home/xl0/work/projects/lovely-tensors
    Traceback (most recent call last):
      File "<string>", line 2, in <module>
      File "<pip-setuptools-caller>", line 34, in <module>
      File "/home/xl0/work/projects/lovely-tensors/setup.py", line 85, in <module>
        setuptools.setup(
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/__init__.py", line 108, in setup
        return distutils.core.setup(**attrs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 184, in setup
        return run_commands(dist)
               ^^^^^^^^^^^^^^^^^^
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 200, in run_commands
        dist.run_commands()
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 964, in run_commands
        self.run_command(cmd)
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/dist.py", line 945, in run_command
        super().run_command(command)
      File "/home/xl0/mambaforge/envs/torch/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 983, in run_command
        cmd_obj.run()
      File "/home/xl0/work/projects/lovely-tensors/setup.py", line 53, in run
        path = os.path.join(self.build_directory, LOVELY_TENSORS_HOOK)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<frozen posixpath>", line 76, in join
    TypeError: expected str, bytes or os.PathLike object, not NoneType
    [end of output]

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by baldassarreFe** [comment]
Hey thanks for testing it!

1. Editable install with `pip install -e .` fails: yes, I tried on my side and fails too. I think it's a matter of getting the correct paths for creating the hook and "installing" it ie copying it to the right place. I'll look into this.

2. In my opinion, there are 3 levels of "magic" for enabling lovely tensors:
  * No magic, just `import lovely_tensors; lovely_tensors.monkey_patch()`, as it is now
  * Auto-patching upon import, ie it's enough to call `import lovely_tensors` and it will immediately call `monkey_patch()`
  * No import required (this PR), just install it and set an env variable, which is neat because it will work out-of-the-box for 3rd party scripts that the user doesn't want to edit

We can get all 3 behaviors but we'll need the import hook and a multi-choice env variable, eg `LOVELY_TENSORS_ENABLED=no`, `LOVELY_TENSORS_ENABLED=upon_import`, `LOVELY_TENSORS_ENABLED=always`.

**Comment by xl0** [comment]
1. Thank you, please let me know if you get stuck, I'll look into it.
2. I played with it some more, and the current implementation works as expected. I think I had some issues with environment variables between conda and Jupyter running in Cursor, which itself is an AppImage on Linux. I'm happy with the current implementation, no further changes required.

**Comment by baldassarreFe** [comment]
Updates:
- I read up more on import hooks here: https://docs.python.org/3/library/site.html
- Took keep things clean, I separated the logic:
  - `_lovely_tensors_hook.pth` is the import hook, its name must end in `.pth` and it can only contain simple one-line instructions
  - `_lovely_tensors_hook.py` contains the monkey patching logic: if `LOVELY_TENSORS=always` then import lovely tensors and monkey patch (delay the import after the check to avoid importing the whole pytorch unecessarily)
- I also modified `__init__.py` to add the other option: if `LOVELY_TENSORS=upon_import` then call `monkey_patch()`

I would be great for you to try it on your machine, just to make sure that it works. The following commands assume that you have the repo at `~/lovely-tensors` and use `uv` to set up a temporary environment.

Setup:
```bash
mkdir -p ~/tmp
TMP_DIR="$(mktemp -p $HOME/tmp/ -d)"
pushd "${TMP_DIR:?}"
uv venv --python 3.12
source .venv/bin/activate
uv pip install torch --index-url https://download.pytorch.org/whl/cpu
uv pip install lovely-numpy
uv pip install --no-cache-dir --verbose ~/lovely-tensors
```

Tests:
```bash
python -c 'import torch; print(torch.ones(3))'
# tensor([1., 1., 1.])

LOVELY_TENSORS= python -c 'import torch; print(torch.ones(3))'
# tensor([1., 1., 1.])

LOVELY_TENSORS=foo python -c 'import torch; print(torch.ones(3))'
# tensor([1., 1., 1.])

LOVELY_TENSORS=always python -c 'import torch; print(torch.ones(3))'
# tensor[3] x∈[1.000, 1.000] μ=1.000 σ=0. [1.000, 1.000, 1.000]

LOVELY_TENSORS=upon_import python -c 'import torch; import lovely_tensors; print(torch.ones(3))'
# tensor[3] x∈[1.000, 1.000] μ=1.000 σ=0. [1.000, 1.000, 1.000]

python -c 'import torch; import lovely_tensors; lovely_tensors.monkey_patch(); print(torch.ones(3))'
# tensor[3] x∈[1.000, 1.000] μ=1.000 σ=0. [1.000, 1.000, 1.000]
```

Clean up:
```bash
deactivate
popd
rm -rf "${TMP_DIR:?}"

rm -rf ~/lovely-tensors/build ~/lovely-tensors/lovely-tensors.egg-info
```

I found that cleaning up the build folders is necessary otherwise pip will skip the build and ignore changes across reinstalls.

Also, I fixed the editable builds so that `uv pip install --no-cache-dir --verbose --editable ~/lovely-tensors` works without errors. However, with an editable build, the `.pth` hook doesn't get called so `LOVELY_TENSORS=upon_import` works but `LOVELY_TENSORS=always` does not. I don't know if this is a limitation of editable builds or a problem with the installation commands. We can dig out more or simply mention this limitation in the docs. I don't think users will do `--editable` anyway.

**Comment by xl0** [comment]
To be honest, I feel the previous version was a better choice.

I see a clear case for having this feature - you have an existing script that prints out some tensors. You set `LOVEY_TENSORS=1`, the tensors now look nice. Same as with "always" now. Same for Notebooks - I was initially confused and thought the .pth was not handled properly, but that was my mistake, things just work.

But if you have a codebase that already imports lovely_tensors, I  think it's a bad idea to make the behavior depend on the variable. For simple `print(t)` it's ok, but `monkey_patch()` adds a dozed new methods to Tensor, and if you use any of them, your code just breaks without this environment variable, likely in the hands of a user that did not have the variable set, and they have no idea what's going on.

Can we just keep it as `if os.environ.get("LOVELY_TENSORS") not in [None, "", "false", "0"]: import and patch`?

Sorry, my earlier comment was not thought through.

2 other points, and sorry for making this review longer.

1. Is there a reason the files start with an underscore? I feel this makes it harder to find them, because they won't show near lovely_tensors in `ls`, and other .pth files in my system don't have underscores.

2. In `setup.py` we have
```
cmdclass = {
        'build': BuildWithHook,
        'easy_install': EasyInstallWithHook,
        'install_lib': InstallLibWithHook,
        'develop': DevelopWithHook,
    },
```

I don't think we need `easry_install` and `develop` - those seem to have been deprecated for a long time.
I don't think we need `build` either, simply copying the files in install_lib just works, right?

**Comment by baldassarreFe** [comment]
> To be honest, I feel the previous version was a better choice.

I agree, just a binary variable that turns the automatic hook on or off is more clear, and more similar to better exceptions.

To summarize, we envision two use cases:
- User is working in a notebook, imports lovely tensors and explicitly calls monkey patch to get all goodies like string, channels, plot etc.
- User is debugging a 3rd party script that crashes mid-run e.g. due to a shape mismatch. However, the stack trace is messed up and not very informative. User installs better exceptions + lovely tensors and sets the corresponding env variables. Without any change to the 3rd party code, the next time the script crashes the stack trace will contain nicely formatted tensors.

I pushed a new commit for this change.

**Comment by baldassarreFe** [comment]
> Is there a reason the files start with an underscore?

Yes, I just wanted to mark the distinction:
- The files under `lovely_tensors/` constitute the library, users are expected to import those files (or the module as a whole) and call methods from it
- The hook files `_lovely_tensors_hook.{py,pth}` are "special" and are not meant to be imported directly by the user

I added the underscore after looking at more examples of import hooks, some of them use the underscore, some others don't:
- RapidsAI: uses an underscore [`_rapids_dask_dependency.pth`](https://github.com/rapidsai/rapids-dask-dependency/blob/branch-24.12/_rapids_dask_dependency.pth)
- Setuptools: no underscore for the hook `distutils-precedence.pth`, but uses an underscore for the module that gets imported in the hook [`_distutils_hack`](https://github.com/pypa/setuptools/tree/main/_distutils_hack)
- Better exceptions: no underscore [`better_exceptions_hook.pth`](https://github.com/Qix-/better-exceptions/blob/master/better_exceptions_hook.pth)

I think the underscore is good to tell the user "don't import this file", but if you don't like it there is no requirement to have it.

**Comment by baldassarreFe** [comment]
> I don't think we need easry_install and develop - those seem to have been deprecated for a long time.

Ok, removed the custom commands for `easy_install` and `develop`. The former is deprecated and the latter does not install the hook properly anyway.

**Comment by xl0** [comment]
@baldassarreFe Thank you! Merged.

**Comment by xl0** [comment]
@baldassarreFe Would you like to add it to the README?

**Comment by baldassarreFe** [comment]
Great! I'll leave the documentation to you if you don't mind. I'm not familiar with the nbdev workflow. Should we also bump the version?

**Comment by xl0** [comment]
Wow, looks really good with BE! I'll document it and bump the version.

**Pull request merged**